### PR TITLE
feat(connectors): add resources

### PR DESCRIPTION
**Pull request by oupo1337**

### CODE REVIEW

**Review by lwintjen — COMMENTED** [review]
@oupo1337 Do you think we could we add a couple of tests in `https://github.com/formalco/terraform-provider-formal/blob/main/tests/main.tf` for the new resources ?

**Comment by oupo1337** [comment]
> @oupo1337 Do you think we could we add a couple of tests in `https://github.com/formalco/terraform-provider-formal/blob/main/tests/main.tf` for the new resources ?

I removed those tests, sorry @lwintjen :(

File: formal/resources/resource_connector_hostname.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+				// This description is used by the documentation generator and the language server.
+				Description: "The ID of this Connector Hostname.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"connector_id": {
+				// This description is used by the documentation generator and the language server.
+				Description: "The ID of the Connector this hostname is linked to.",
+				Type:        schema.TypeString,
+				Required:    true,
```
Comment (ahmb84) [inline formal/resources/resource_connector_hostname.go:44]: add ForceNew: true

File: formal/resources/resource_connector_hostname.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+			},
+			"hostname": {
+				// This description is used by the documentation generator and the language server.
+				Description: "The hostname for this Connector hostname.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
```
Comment (ahmb84) [inline formal/resources/resource_connector_hostname.go:51]: add ForceNew: true

File: formal/resources/resource_connector_hostname.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+func resourceConnectorHostnameUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	// use the meta value to retrieve your client from the provider configure method
+	c := meta.(*clients.Clients)
+
+	// Warning or errors can be collected in a slice type
+	var diags diag.Diagnostics
+
+	connectorHostnameId := d.Id()
+
+	fieldsThatCanChange := []string{"connector_id", "hostname", "managed_tls", "termination_protection"}
```
Comment (ahmb84) [inline formal/resources/resource_connector_hostname.go:134]: connector_id, and hostname cannot change

File: formal/resources/resource_connector_listener_link.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+				// This description is used by the documentation generator and the language server.
+				Description: "The ID of this connector listener link.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"connector_listener_id": {
+				// This description is used by the documentation generator and the language server.
+				Description: "The ID of the Connector Listener you want to link to a connector.",
+				Type:        schema.TypeString,
+				Required:    true,
```
Comment (ahmb84) [inline formal/resources/resource_connector_listener_link.go:44]: add ForceNew: true

File: formal/resources/resource_connector_listener_link.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+			},
+			"connector_id": {
+				// This description is used by the documentation generator and the language server.
+				Description: "The ID of the Connector Listener you want to link to a connector.",
+				Type:        schema.TypeString,
+				Required:    true,
```
Comment (ahmb84) [inline formal/resources/resource_connector_listener_link.go:50]: add ForceNew: true

File: formal/resources/resource_connector_listener_link.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+func resourceConnectorListenerLinkUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	// use the meta value to retrieve your client from the provider configure method
+	c := meta.(*clients.Clients)
+
+	// Warning or errors can be collected in a slice type
+	var diags diag.Diagnostics
+
+	connectorListenerLinkId := d.Id()
+
+	fieldsThatCanChange := []string{"connector_listener_id", "connector_id", "termination_protection"}
```
Comment (ahmb84) [inline formal/resources/resource_connector_listener_link.go:129]: connector_listener_id", "connector_id", can't be updated

**Comment by ahmb84** [comment]
For every fields that are "updatable" make sure that it is the case at the api layer

**Review by ahmb84 — CHANGES_REQUESTED** [review]

File: formal/resources/resource_connector.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	resApiKey, err := c.Grpc.Sdk.ConnectorServiceClient.GetConnectorApiKey(ctx, connect.NewRequest(&corev1.GetConnectorApiKeyRequest{Id: connectorId}))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.Set("id", res.Msg.Connector.Id)
+	d.Set("name", res.Msg.Connector.Name)
+	d.Set("api_key", resApiKey.Msg.Secret)
```
Comment (lwintjen) [inline formal/resources/resource_connector.go:106]: I believe this does not exist in the scema, is it a leftover?

Comment (oupo1337) [reply formal/resources/resource_connector.go:113]: fixed 👌

Comment (ahmb84) [reply formal/resources/resource_connector.go:113]: it should exist, a connector has an api key

**Pull request merged**